[cli] ensure that development build is installed when opening interstitial page

esamelson opened the pull request:
# Why

second step of ENG-6210 -- port https://github.com/expo/expo-cli/commit/f1624c75b52cc1c4f99354ec4021494e0eff74aa over to new CLI

# How

- Added BundlerDevServer.getInterstitialPageUrl since PlatformManager may now decide to use the Expo Go URL even if the interstitial page URL is nonnull.
- Added PlatformManager.openProjectInterstitialPageAsync to keep the interstitial logic separate from openProjectInExpoGoAsync. Now `openAsync` determines which of these two methods to call in the case of `runtime === 'expo'`.
- Also updated an older error message to use the modern terminology of "development build" instead of "development client"

# Test Plan

Added & updated tests for new behavior.
Also manually tested the following scenarios to ensure parity with old CLI:
- `EXPO_ENABLE_INTERSTITIAL_PAGE=1 npx expo start` + i/a with both dev build and Expo Go installed
- `EXPO_ENABLE_INTERSTITIAL_PAGE=1 npx expo start` + i/a with dev build but no Expo Go
- `EXPO_ENABLE_INTERSTITIAL_PAGE=1 npx expo start` + i/a with Expo Go but no dev build
- `EXPO_ENABLE_INTERSTITIAL_PAGE=1 npx expo start` + i/a with neither dev build nor Expo Go
- `EXPO_ENABLE_INTERSTITIAL_PAGE=1 npx expo start` + i/a without expo-dev-client installed in project
- `npx expo start --dev-client` + i/a without expo-dev-client installed in project

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

esamelson committed:
[cli] ensure that development build is installed when opening interstitial page

EvanBacon commented on packages/@expo/cli/CHANGELOG.md:
```diff
@@ -26,6 +26,7 @@
 - Fix `prebuild` e2e tests. ([#18612](https://github.com/expo/expo/pull/18612) by [@EvanBacon](https://github.com/EvanBacon))
 - Add warning about malformed project when running prebuild in non-interactive mode. ([#18436](https://github.com/expo/expo/pull/18436) by [@wkozyra95](https://github.com/wkozyra95))
 - [Interstitial page] Capture missing analytics event when user opens development build. ([#18792](https://github.com/expo/expo/pull/18792) by [@esamelson](https://github.com/esamelson))
+- [Interstitial page] Ensure that development build is installed. ([#18836](https://github.com/expo/expo/pull/18836) by [@esamelson](https://github.com/esamelson))
```
```suggestion
- [Interstitial page] Ensure that the development build is installed when opening the interstitial page. ([#18836](https://github.com/expo/expo/pull/18836) by [@esamelson](https://github.com/esamelson))
```

EvanBacon commented on packages/@expo/cli/src/start/platforms/PlatformManager.ts:
```diff
@@ -37,6 +38,8 @@ export class PlatformManager<
       /** Dev Client */
       getCustomRuntimeUrl: (props?: { scheme?: string }) => string | null;
+      /** Interstitial page URL, or null if the interstitial page should not open */
+      getInterstitialPageUrl: () => string | null;
```
There is already interstitial logic in the `getExpoGoUrl` function.

EvanBacon commented on packages/@expo/cli/src/start/platforms/PlatformManager.ts:
```diff
@@ -74,6 +76,31 @@ export class PlatformManager<
+  private async openProjectInterstitialPageAsync(
+    deviceManager: DeviceManager<IDevice>
+  ): Promise<{ url: string }> {
+    const url = this.props.getInterstitialPageUrl();
+    // This should never happen, but just in case...
+    assert(url, 'Could not get interstitial page URL');
+
+    deviceManager.logOpeningUrl(url);
+
+    // TODO: Expensive, we should only do this once.
+    const { exp } = getConfig(this.projectRoot);
+    const installedExpo = await deviceManager.ensureExpoGoAsync(exp.sdkVersion);
+
+    deviceManager.activateWindowAsync();
+    await deviceManager.openUrlAsync(url);
+
+    // TODO: probably don't want to log this here, but keeping it for parity with old CLI
+    await logEventAsync('Open Url on Device', {
+      platform: this.props.platform,
+      installedExpo,
+    });
+
+    return { url };
```
No copy/paste complex functions, especially with two TODOs inside them.

EvanBacon commented on packages/@expo/cli/src/start/platforms/PlatformManager.ts:
```diff
@@ -136,7 +163,22 @@ export class PlatformManager<
     );
     if (options.runtime === 'expo') {
-      return this.openProjectInExpoGoAsync(resolveSettings);
+      const deviceManager = await this.props.resolveDeviceAsync(resolveSettings);
+      if (this.props.getInterstitialPageUrl()) {
+        const applicationId = await this._getAppIdResolver().getAppIdAsync();
+        if (await deviceManager.isAppInstalledAsync(applicationId)) {
+          return this.openProjectInterstitialPageAsync(deviceManager);
+        } else {
+          Log.warn(
+            `\u203A The 'expo-dev-client' package is installed, but a development build isn't ` +
+              `available.\nYour app will open in Expo Go instead. If you want to use a ` +
+              `development build, you need to make and install one first.\n${learnMore(
+                'https://docs.expo.dev/development/build/'
+              )}`
+          );
+        }
+      }
+      return this.openProjectInExpoGoAsync(deviceManager);
```
The `openAsync` function currently delegates out complex open options to preserve sanity and keep the code readable, this change seems to extract the existing functionality from `openProjectInExpoGoAsync` and `getExpoGoUrl` into `openAsync`.

EvanBacon reviewed (changes requested):
This PR looks like it reimplements all of the existing functionality but in a less organized way, probably shouldn't be merged.

EvanBacon commented on packages/@expo/cli/src/start/platforms/PlatformManager.ts:
```diff
@@ -136,7 +163,22 @@ export class PlatformManager<
+        if (await deviceManager.isAppInstalledAsync(applicationId)) {
+          return this.openProjectInterstitialPageAsync(deviceManager);
+        } else {
+          Log.warn(
+            `\u203A The 'expo-dev-client' package is installed, but a development build isn't ` +
+              `available.\nYour app will open in Expo Go instead. If you want to use a ` +
+              `development build, you need to make and install one first.\n${learnMore(
+                'https://docs.expo.dev/development/build/'
+              )}`
+          );
```
I imagine this would go here: [assertion](https://github.com/expo/expo/blob/9a148a31a58007f230f4f6b2369811c2649bf989/packages/%40expo/cli/src/start/platforms/PlatformManager.ts#L57)?

esamelson replied:
Since this log message is specific to the case where no dev build is installed on the device, but the interstitial page _would_ otherwise be opened, it has to live in `shouldUseInterstitialPage` now that the logic has moved there too.

esamelson committed:
move logic into openProjectInExpoGoAsync

esamelson commented:
> This PR looks like it reimplements all of the existing functionality but in a less organized way, probably shouldn't be merged.

This PR adds an additional check to see if a development build is installed on the device before opening the interstitial page. (Added to expo-cli in https://github.com/expo/expo-cli/commit/f1624c75b52cc1c4f99354ec4021494e0eff74aa.) Currently this CLI only checks if expo-dev-client is installed in the project.

I thought that it would be more organized to have separate functions for the individual cases, and tried to keep all of the calls to `deviceManager` in this one class, but I'm happy to change it given that you have a different opinion. I've moved the new check into `openProjectInExpoGoAsync` and renamed it to `openProjectInExpoGoOrInterstitialPageAsync` to be clearer (that's what it already does anyway). I could instead move the new check even further up into BundlerDevServer if you'd prefer.

EvanBacon commented on packages/@expo/cli/src/start/platforms/PlatformManager.ts:
```diff
@@ -49,14 +52,31 @@ export class PlatformManager<
+    const deviceManager = await this.props.resolveDeviceAsync(resolveSettings);
+
+    let url = expoGoUrl;
+    if (interstitialPageUrl) {
```
I'd prefer if this was the `shouldUseInterstitialPage()` and the URL logic stays combined. The model here is spiritually similar to universal links on iOS: user open apps (Expo Go) -> more than one target -> disambiguate. But the intention should always be "open in dev client" or "open in Expo Go" instead of "open in interstitial".

EvanBacon commented on packages/@expo/cli/src/start/platforms/PlatformManager.ts:
```diff
@@ -136,7 +156,7 @@ export class PlatformManager<
     );
     if (options.runtime === 'expo') {
-      return this.openProjectInExpoGoAsync(resolveSettings);
+      return this.openProjectInExpoGoOrInterstitialPageAsync(resolveSettings);
```
```suggestion
      return this. openProjectInExpoGoAsync(resolveSettings);
```

I'd prefer this name stay the same for the [reasoning here](https://github.com/expo/expo/pull/18836/files#r966079250).

esamelson committed:
feedback

esamelson committed:
move logic into shouldUseInterstitialPage

esamelson committed:
fix lint warnings

esamelson committed:
move side effect out of shouldUseInterstitialPage for #18838

esamelson committed:
[cli] point QR code to interstitial page when enabled (#18838)

EvanBacon committed:
Merge branch 'main' into @eric/cli-ensure-dc-installed

EvanBacon committed:
Update expo-cli.md
